File: zero-server/src/main/kotlin/com/zeroday/service/ServerScheduler.kt
package com.zeroday.service

import com.zeroday.handler.ConnectionRegistry
import kotlinx.coroutines.*
import org.slf4j.LoggerFactory

/**
 * Periodic server-wide tasks that run in the background: stats logging
 * and maintenance cleanup.
 *
 * Tasks run in a single loop every 60 s. Each task's [block] is wrapped
 * in a try-catch so a single failing task cannot stall the scheduler.
 */
class ServerScheduler(
    private val playerService: PlayerService,
    private val connectionRegistry: ConnectionRegistry,
    private val persistence: PlayerPersistence? = null
) {
    private val log = LoggerFactory.getLogger(ServerScheduler::class.java)
    private var job: Job? = null

    /**
     * Start the background loop. Replaces any previously running loop.
     */
    fun start(scope: CoroutineScope) {
        stop()
        job = scope.launch {
            while (isActive) {
                try {
                    val online = playerService.getOnlinePlayers().size
                    val total = playerService.getAllPlayers().size
                    val conns = connectionRegistry.totalConnections()
                    if (online > 0 || conns > 0) {
                        log.info("scheduler stats: online={} total={} connections={}", online, total, conns)
                    }
                } catch (e: Exception) {
                    log.warn("scheduler task failed", e)
                }
                delay(60_000L)
            }
        }
    }

    fun stop() {
        job?.cancel()
        job = null
    }
}

File: zero-server-csharp/src/ZeroDayMMO.Server/Services/ServerScheduler.cs
using Microsoft.Extensions.Hosting;
using Microsoft.Extensions.Logging;

namespace ZeroDayMMO.Server.Services;

public class ServerScheduler : IHostedService, IDisposable
{
    private readonly PlayerService _playerService;
    private readonly ConnectionRegistry _connectionRegistry;
    private readonly PlayerPersistence? _persistence;
    private readonly ILogger<ServerScheduler> _logger;
    private Timer? _timer;

    public ServerScheduler(
        PlayerService playerService,
        ConnectionRegistry connectionRegistry,
        PlayerPersistence? persistence,
        ILogger<ServerScheduler> logger)
    {
        _playerService = playerService;
        _connectionRegistry = connectionRegistry;
        _persistence = persistence;
        _logger = logger;
    }

    public Task StartAsync(CancellationToken cancellationToken)
    {
        _timer = new Timer(DoWork, null, TimeSpan.FromSeconds(60), TimeSpan.FromSeconds(60));
        return Task.CompletedTask;
    }

    public Task StopAsync(CancellationToken cancellationToken)
    {
        _timer?.Change(Timeout.Infinite, 0);
        return Task.CompletedTask;
    }

    private void DoWork(object? state)
    {
        try
        {
            var online = _playerService.GetOnlinePlayers().Count;
            var total = _playerService.GetAllPlayers().Count();
            var conns = _connectionRegistry.ActiveConnections;

            if (online > 0 || conns > 0)
            {
                _logger.LogInformation("scheduler stats: online={Online} total={Total} connections={Conns}", online, total, conns);
            }
        }
        catch (Exception ex)
        {
            _logger.LogWarning(ex, "scheduler task failed");
        }
    }

    public void Dispose()
    {
        _timer?.Dispose();
    }
}
